src/engine/tests/Test_CertificateNode.h

```diff
@@ -0,0 +1,200 @@
+        TS_ASSERT_EQUALS( root.getPLCExplanations().size(), ( unsigned ) 3 );
+
+        root.removePLCExplanationsBelowDecisionLevel( 0 );
+        TS_ASSERT_EQUALS( root.getPLCExplanations().size(), ( unsigned ) 2 );
+        TS_ASSERT_EQUALS( root.getPLCExplanations().front(), explanation2 );
+        TS_ASSERT_EQUALS( root.getPLCExplanations().back(), explanation3 );
+
+        root.resizePLCExplanationsList( 1 );
+        TS_ASSERT_EQUALS( root.getPLCExplanations().size(), ( unsigned ) 1 );
+        TS_ASSERT_EQUALS( root.getPLCExplanations().back(), explanation2 );
+
+        root.deletePLCExplanations();
+        TS_ASSERT( root.getPLCExplanations().empty() );
+
+        // removePLCExplanationsBelowDecisionLevel doesn't delete by itself
+        delete explanation1;
+    }
+
+    /*
+       Tests certification methods
+     */
+    void testCertification()
+    {
+        Vector<double> row1 = { 1, 0, -1, 0, 1, 0, 0 }; // Row of ReLU1
+        Vector<double> row2 = { 0, 1, 0, -1, 0, 1, 0 }; // Row of ReLU2
+        Vector<double> row3 = { 0.5, 0, -1, 0, 0, 0, 1 };
+        Vector<Vector<double>> initialTableau = { row1, row2, row3 };
+
+        Vector<double> groundUpperBounds( row1.size(), 1 );
+        Vector<double> groundLowerBounds( row1.size(), 0 );
+        groundUpperBounds[6] = 2;
+
+        // Set a complete tree of depth 3, using 2 ReLUs
+        auto *root = new CertificateNode( &initialTableau, groundUpperBounds, groundLowerBounds );
+
+        ReluConstraint relu1 = ReluConstraint( 0, 2 ); // aux var is 4
+        ReluConstraint relu2 = ReluConstraint( 1, 3) ; // aux var is 5
+
+        root->addProblemConstraint( RELU, { 0, 2, 4 }, PHASE_NOT_FIXED );
+        root->addProblemConstraint( RELU, { 1, 3, 5 }, PHASE_NOT_FIXED );
+
+        auto splits1 = relu1.getCaseSplits();
+        auto splits2 = relu2.getCaseSplits();
+        TS_ASSERT_EQUALS( splits1.size(), ( unsigned ) 2 );
+        TS_ASSERT_EQUALS( splits2.size(), ( unsigned ) 2 );
+
+        PiecewiseLinearCaseSplit split1_1 = splits1.back();
+        PiecewiseLinearCaseSplit split1_2 = splits1.front();
+
+        PiecewiseLinearCaseSplit split2_1 = splits2.back();
+        PiecewiseLinearCaseSplit split2_2 = splits2.front();
+
+        TS_ASSERT_EQUALS( split1_2.getBoundTightenings().size(), ( unsigned ) 2 );
+
+        auto *child1 = new CertificateNode( root, split1_1 ); // Child with missing aux tightening
+        auto *child2 = new CertificateNode( root, split1_2 );
+
+        auto *child2_1 = new CertificateNode( child2, split2_1 );
+        auto *child2_2 = new CertificateNode( child2, split2_2 );
+
+        root->wasVisited();
+        child2->wasVisited();
+        child1->wasVisited();
+        child2_1->wasVisited();
+
+        // Visited leaves have no contradiction, so certification will fail
+        TS_ASSERT( !root->certify() );
+
+        // Mark visited leaves with flags that immediately certify them
+        child1->hasSATSolution();
```

Comment:
I don't understand. If child 1 is SAT, why does certify() succeed?

Reply:
This is a part of the mechanism that allows checking the certificate for SAT queries as well (so we have more test cases). Certification in this case checks that all UNSAT leaves have a correct contradiction, and all others (not searched, are SAT, etc.) are skipped.
Conceptually, certify() succeeds since everything that has a proof is correctly proven.